aptos-move/framework/src/natives/cryptography/algebra/mod.rs

```diff
@@ -0,0 +1,319 @@
+            abort_unless_feature_enabled!(context, FeatureFlag::BLS12_381_STRUCTURES);
+            let buf = ark_serialize_internal!(
+                gas_params,
+                context,
+                Structure::BLS12381Fr,
+                handle,
+                SerializationFormat::BLS12381FrLsb,
+                ark_bls12_381::Fr,
+                serialize_uncompressed
+            );
+            Ok(smallvec![Value::vector_u8(buf)])
+        },
+        _ => Err(SafeNativeError::Abort {
+            abort_code: MOVE_ABORT_CODE_NOT_IMPLEMENTED,
+        }),
+    }
+}
+
+macro_rules! ark_deserialize_internal {
+    (
+        $gas_params:expr,
+        $context:expr,
+        $bytes:expr,
+        $structure:expr,
+        $format:expr,
+        $typ:ty,
+        $deser_func:ident
+    ) => {{
+        $context.charge($gas_params.placeholder)?;
+        match <$typ>::$deser_func($bytes) {
+            Ok(element) => {
+                let handle = store_obj!($context, element);
+                Ok(smallvec![Value::bool(true), Value::u64(handle as u64)])
+            },
+            _ => Ok(smallvec![Value::bool(false), Value::u64(0)]),
+        }
+    }};
+}
+
+fn deserialize_internal(
+    gas_params: &GasParameters,
+    context: &mut SafeNativeContext,
+    ty_args: Vec<Type>,
+    mut args: VecDeque<Value>,
+) -> SafeNativeResult<SmallVec<[Value; 1]>> {
+    assert_eq!(1, ty_args.len());
+    let structure = structure_from_ty_arg!(context, &ty_args[0]);
+    let vector_ref = safely_pop_arg!(args, VectorRef);
+    let bytes_ref = vector_ref.as_bytes_ref();
+    let bytes = bytes_ref.as_slice();
+    let format_opt = SerializationFormat::try_from(safely_pop_arg!(args, u64));
+    match (structure, format_opt) {
+        (Some(Structure::BLS12381Fr), Ok(SerializationFormat::BLS12381FrLsb)) => {
+            if bytes.len() != 32 {
+                return Ok(smallvec![Value::bool(false), Value::u64(0)]);
+            }
+            ark_deserialize_internal!(
+                gas_params,
+                context,
+                bytes,
+                Structure::BLS12381Fr,
+                SerializationFormat::BLS12381FrLsb,
+                ark_bls12_381::Fr,
+                deserialize_uncompressed
+            )
+        },
+        _ => Err(SafeNativeError::Abort {
+            abort_code: MOVE_ABORT_CODE_NOT_IMPLEMENTED,
+        }),
+    }
+}
+
+macro_rules! ark_field_add_internal {
+    ($gas_params:expr, $context:expr, $args:ident, $structure:expr, $typ:ty) => {{
+        let handle_2 = safely_pop_arg!($args, u64) as usize;
+        let handle_1 = safely_pop_arg!($args, u64) as usize;
+        let element_1_ptr = get_obj_pointer!($context, handle_1);
+        let element_1 = element_1_ptr.downcast_ref::<$typ>().unwrap();
+        let element_2_ptr = get_obj_pointer!($context, handle_2);
+        let element_2 = element_2_ptr.downcast_ref::<$typ>().unwrap();
```

Comment:
If unwrap() fails here... we are toast... we need to handle this properly & abort.

Reply:
Keeping one thread open for the same issue.